File: android/app/src/main/java/com/bayesmech/vision/analysis/InsightRepository.kt
package com.bayesmech.vision.analysis

import android.content.Context
import com.bayesmech.vision.ChatHistory
import com.bayesmech.vision.ChatTurn
import com.bayesmech.vision.GensparkSummary
import com.bayesmech.vision.InsightVideoResponse
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

data class ChatSyncResult(
    val history: ChatHistory?,
    val newTurns: List<ChatTurn>,
)

data class FollowUpResult(
    val sessionId: String?,
    val responseText: String,
    val userTimestampNs: Long,
    val responseTimestampNs: Long,
)

class InsightRepository(
    context: Context,
    private val httpClient: OkHttpClient = OkHttpClient.Builder()
        .readTimeout(120, TimeUnit.SECONDS)
        .build(),
) {

    private val cacheStore = InsightCacheStore(context)

    suspend fun getSummary(
        serverUrl: String,
        fileName: String,
        forceRefresh: Boolean = false,
    ): GensparkSummary? = withContext(Dispatchers.IO) {
        val cached = cacheStore.readSummary(fileName)
        if (!forceRefresh && cached != null) {
            return@withContext cached
        }

        val remote = runCatching {
            val request = Request.Builder()
                .url("${httpBase(serverUrl)}/api/insightgen/insight?file=$fileName")
                .get()
                .build()
            httpClient.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return@runCatching null
                val bytes = response.body?.bytes() ?: return@runCatching null
                GensparkSummary.parseFrom(bytes)
            }
        }.getOrNull()

        remote?.let { cacheStore.writeSummary(fileName, it) }
        remote ?: cached
    }

    suspend fun getVideoLayer(
        serverUrl: String,
        fileName: String,
        layerName: String,
        forceRefresh: Boolean = false,
    ): InsightVideoResponse? = withContext(Dispatchers.IO) {
        val cached = cacheStore.readVideoLayer(fileName, layerName)
        if (!forceRefresh && cached != null) {
            return@withContext cached
        }

        val remote = runCatching {
            val request = Request.Builder()
                .url("${httpBase(serverUrl)}/api/insightgen/video?file=$fileName&layer=$layerName")
                .get()
                .build()
            httpClient.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return@runCatching null
                val bytes = response.body?.bytes() ?: return@runCatching null
                InsightVideoResponse.parseFrom(bytes)
            }
        }.getOrNull()

        remote?.let { cacheStore.writeVideoLayer(fileName, layerName, it) }
        remote ?: cached
    }

    suspend fun readCachedChatHistory(fileName: String): ChatHistory? = withContext(Dispatchers.IO) {
        cacheStore.readChatHistory(fileName)
    }

    suspend fun syncChatHistory(serverUrl: String, fileName: String): ChatSyncResult =
        withContext(Dispatchers.IO) {
            val cached = cacheStore.readChatHistory(fileName)
            val sinceTimestampNs = cacheStore.latestMessageTimestampNs(fileName)

            val remoteDelta = runCatching {
                val request = Request.Builder()
                    .url("${httpBase(serverUrl)}/api/insightgen/chat?file=$fileName&since_timestamp_ns=$sinceTimestampNs")
                    .get()
                    .build()
                httpClient.newCall(request).execute().use { response ->
                    if (!response.isSuccessful) return@runCatching null
                    val bytes = response.body?.bytes() ?: return@runCatching null
                    ChatHistory.parseFrom(bytes)
                }
            }.getOrNull()

            if (remoteDelta == null) {
                return@withContext ChatSyncResult(history = cached, newTurns = emptyList())
            }

            val merged = cacheStore.mergeChatHistory(fileName, remoteDelta)
            ChatSyncResult(history = merged, newTurns = remoteDelta.turnsList)
        }

    suspend fun sendFollowUp(
        serverUrl: String,
        fileName: String,
        message: String,
        sessionId: String?,
    ): FollowUpResult? = withContext(Dispatchers.IO) {
        runCatching {
            val body = JSONObject().apply {
                put("file", fileName)
                put("message", message)
                sessionId?.let { put("session_id", it) }
            }.toString().toRequestBody("application/json".toMediaType())

            val request = Request.Builder()
                .url("${httpBase(serverUrl)}/api/insightgen/chat")
                .post(body)
                .build()

            httpClient.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return@runCatching null
                val payload = JSONObject(response.body?.string() ?: return@runCatching null)
                FollowUpResult(
                    sessionId = payload.optString("session_id").takeIf { it.isNotBlank() } ?: sessionId,
                    responseText = payload.optString("response"),
                    userTimestampNs = payload.optLong("user_timestamp_ns"),
                    responseTimestampNs = payload.optLong("response_timestamp_ns"),
                )
            }
        }.getOrNull()
    }

    suspend fun cacheChatExchange(
        fileName: String,
        userMessage: String,
        responseText: String,
        userTimestampNs: Long,
        responseTimestampNs: Long,
    ) = withContext(Dispatchers.IO) {
        cacheStore.appendChatTurns(
            fileName = fileName,
            turns = listOf(
                ChatTurn.newBuilder()
                    .setRole("user")
                    .setText(userMessage)
                    .setTimestampNs(userTimestampNs)
                    .build(),
                ChatTurn.newBuilder()
                    .setRole("model")
                    .setText(responseText)
                    .setTimestampNs(responseTimestampNs)
                    .build(),
            ),
        )
    }

    private fun httpBase(serverUrl: String): String =
        serverUrl.trimEnd('/')
            .replaceFirst("wss://", "https://")
            .replaceFirst("ws://", "http://")
}

File: ios/BayesMechVision/Analysis/InsightRepository.swift
import Foundation
import SwiftProtobuf

struct ChatSyncResult {
    var history: ChatHistory?
    var newTurns: [ChatTurn]
}

struct FollowUpResult {
    var sessionID: String?
    var responseText: String
    var userTimestampNs: Int64
    var responseTimestampNs: Int64
}

final class InsightRepository {
    private let session: URLSession
    private let cacheStore: InsightCacheStore

    init(session: URLSession = .shared, cacheStore: InsightCacheStore = InsightCacheStore()) {
        self.session = session
        self.cacheStore = cacheStore
    }

    func listRecordings(serverURL: String, user: SignedInUser?) async throws -> [DataList] {
        guard let url = URL(string: "\(HTTPBase.from(serverURL: serverURL))/api/insightgen/recordings") else {
            throw URLError(.badURL)
        }

        var requestProto = ListRecordingsRequest()
        let username = [user?.displayName, user?.email]
            .compactMap { $0?.trimmingCharacters(in: .whitespacesAndNewlines) }
            .first(where: { !$0.isEmpty }) ?? "unknown"
        requestProto.username = username
        requestProto.authToken = user?.authToken ?? ""

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-protobuf", forHTTPHeaderField: "Content-Type")
        request.httpBody = try requestProto.serializedData()

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, (200..<300).contains(httpResponse.statusCode) else {
            throw URLError(.badServerResponse)
        }

        return try ListRecordingsResponse(serializedBytes: data).recordings
    }

    func getSummary(serverURL: String, fileName: String, forceRefresh: Bool = false) async -> GensparkSummary? {
        if !forceRefresh, let cached = cacheStore.readSummary(fileName: fileName) {
            return cached
        }

        guard let url = URL(string: "\(HTTPBase.from(serverURL: serverURL))/api/insightgen/insight?file=\(fileName)") else {
            return cacheStore.readSummary(fileName: fileName)
        }

        do {
            let (data, response) = try await session.data(from: url)
            guard let httpResponse = response as? HTTPURLResponse, (200..<300).contains(httpResponse.statusCode) else {
                return cacheStore.readSummary(fileName: fileName)
            }
            let summary = try GensparkSummary(serializedBytes: data)
            cacheStore.writeSummary(fileName: fileName, summary: summary)
            return summary
        } catch {
            return cacheStore.readSummary(fileName: fileName)
        }
    }

    func getVideoLayer(serverURL: String, fileName: String, layerName: String, forceRefresh: Bool = false) async -> InsightVideoResponse? {
        if !forceRefresh, let cached = cacheStore.readVideoLayer(fileName: fileName, layerName: layerName) {
            return cached
        }

        guard let url = URL(string: "\(HTTPBase.from(serverURL: serverURL))/api/insightgen/video?file=\(fileName)&layer=\(layerName)") else {
            return cacheStore.readVideoLayer(fileName: fileName, layerName: layerName)
        }

        do {
            let (data, response) = try await session.data(from: url)
            guard let httpResponse = response as? HTTPURLResponse, (200..<300).contains(httpResponse.statusCode) else {
                return cacheStore.readVideoLayer(fileName: fileName, layerName: layerName)
            }
            let payload = try InsightVideoResponse(serializedBytes: data)
            cacheStore.writeVideoLayer(fileName: fileName, layerName: layerName, response: payload)
            return payload
        } catch {
            return cacheStore.readVideoLayer(fileName: fileName, layerName: layerName)
        }
    }

    func readCachedChatHistory(fileName: String) -> ChatHistory? {
        cacheStore.readChatHistory(fileName: fileName)
    }

    func syncChatHistory(serverURL: String, fileName: String) async -> ChatSyncResult {
        let sinceTimestamp = cacheStore.latestMessageTimestampNs(fileName: fileName)
        guard let url = URL(string: "\(HTTPBase.from(serverURL: serverURL))/api/insightgen/chat?file=\(fileName)&since_timestamp_ns=\(sinceTimestamp)") else {
            return ChatSyncResult(history: cacheStore.readChatHistory(fileName: fileName), newTurns: [])
        }

        do {
            let (data, response) = try await session.data(from: url)
            guard let httpResponse = response as? HTTPURLResponse, (200..<300).contains(httpResponse.statusCode) else {
                return ChatSyncResult(history: cacheStore.readChatHistory(fileName: fileName), newTurns: [])
            }
            let remoteDelta = try ChatHistory(serializedBytes: data)
            let merged = cacheStore.mergeChatHistory(fileName: fileName, remoteDelta: remoteDelta)
            return ChatSyncResult(history: merged, newTurns: remoteDelta.turns)
        } catch {
            return ChatSyncResult(history: cacheStore.readChatHistory(fileName: fileName), newTurns: [])
        }
    }

    func sendFollowUp(serverURL: String, fileName: String, message: String, sessionID: String?) async -> FollowUpResult? {
        guard let url = URL(string: "\(HTTPBase.from(serverURL: serverURL))/api/insightgen/chat") else {
            return nil
        }

        let payload: [String: Any] = [
            "file": fileName,
            "message": message,
            "session_id": sessionID as Any
        ]

        guard let body = try? JSONSerialization.data(withJSONObject: payload) else { return nil }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = body

        do {
            let (data, response) = try await session.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse, (200..<300).contains(httpResponse.statusCode),
                  let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                return nil
            }

            return FollowUpResult(
                sessionID: (json["session_id"] as? String).flatMap { $0.isEmpty ? sessionID : $0 },
                responseText: json["response"] as? String ?? "",
                userTimestampNs: json["user_timestamp_ns"] as? Int64 ?? Int64((json["user_timestamp_ns"] as? NSNumber)?.int64Value ?? 0),
                responseTimestampNs: json["response_timestamp_ns"] as? Int64 ?? Int64((json["response_timestamp_ns"] as? NSNumber)?.int64Value ?? 0)
            )
        } catch {
            return nil
        }
    }

    func cacheChatExchange(fileName: String, userMessage: String, responseText: String, userTimestampNs: Int64, responseTimestampNs: Int64) {
        var userTurn = ChatTurn()
        userTurn.role = "user"
        userTurn.text = userMessage
        userTurn.timestampNs = userTimestampNs

        var modelTurn = ChatTurn()
        modelTurn.role = "model"
        modelTurn.text = responseText
        modelTurn.timestampNs = responseTimestampNs

        cacheStore.appendChatTurns(fileName: fileName, turns: [userTurn, modelTurn])
    }
}
